_posts/2017-07-26-property-based-testing-in-javascript.md

```diff
@@ -0,0 +1,191 @@
+(which is not surprising as they all are based on *Haskells Quickcheck*). To emphasize this here are the definitions
+for a property in `jsverify` and `quick_check` (where property refers to a single element of the array above):
+
+```javascript
+// jsverify
+jv.property(
+    property.description,
+    // First input: A function that returns a boolean for any input
+    jv.fn(jv.bool),
+    // Second input: An array of any values
+    jv.array(any),
+    // The property
+    property.test
+);
+
+// quick_check
+it(property.description, function () {
+    qc.forAll(
+        // First input: A function that returns a boolean for any input
+        qc.function(qc.bool),
+        // Second input: An array of any values
+        qc.array,
+        // The property
+        (...args) => assert(property.test(...args))
+    );
+});
+```
+
+Basically they look the same except for the property description (because jsverify has built-in mocha integration) and
+`quick_check` actually requiring an assert. All frameworks fulfill the most important points for property based testing:
+
+- Integrates well with test frameworks, such as mocha, jasmine, ava, etc.
+- Shrinking: Reduces a failing testcase to the minimum failing test case (e.g. if your test fails for an input array,
+  the test runner will reduce this array to the minimum number of elements to get the failing test)
+- Miscellanious Generators: Numbers, Lists, Objects, Custom and more
+
+I had some issues with the `testcheck` library though. It was horribly slow compared to the other libraries, especially when
+increasing the number of examples that are tested in a single test run. E.g. when testing with 100 examples, the tests
+took almost 20 seconds to finish. Plus, it provides the least amount of pre-defined generators. For our test case I
+especially missed any generator that generates functions, so I had to write my own.
+
+On the other hand, writing a custom generator is not that complicated (for all frameworks). For our function, we need a
+deterministic function that gets a single argument of any type and returns a boolean:
+
+```javascript
+const objectHash = require('object-hash');
+const genFunc = gen.intWithin(0, 9).then((i) => {
+    return function hashEq(val) {
+        return objectHash(val).startsWith(i.toString());
+    }
+});
+```
+
+When a value for this generator is built, we basically get a function that checks whether the hash of the value starts
+with `i`, where `i` is a value drawn from the range from `0` to `9`. This function is deterministic, so we can use it
+as the input for our property test. This concept of generator transformation exists in all of the mentioned frameworks
+(although it is called `map` in the others).
+
+## Conclusion
+
+Property-based testing is a cool and useful addition to your test suite. We showed you how to build your own property-based
```

Comment:
Replace `we showed you` with i.e. `we saw`, because it's basically you, who showed us ;)